html2text.1
.\" manpage for html2text
.\"
.TH html2text 1 2020\-04\-15
.SH NAME
html2text \- an advanced HTML\-to\-text converter
.SH SYNOPSIS
.B html2text -help
.br
.B html2text -version
.br
.B html2text
[
.B \-check
] [
.B \-debug\-scanner
] [
.B \-debug\-parser
] [
.B \-rcfile
.I path
] [
.B \-width
.I width
] [
.B \-o
.I output-file
] [
.B \-nobs
] [
.B \-from_encoding
.I encoding
] [
.B \-to_encoding
.I encoding
] [
.B \-ascii
] [
.B \-utf8
] [
.IR input-file " ..."
]
.SH DESCRIPTION
.B html2text
reads HTML documents from the
.IR input-file s,
formats each of them into a stream of UTF-8 encoded characters,
and writes the result to standard output (or into
.IR output-file ,
if the
.B -o
command line option is used).
.P
If no
.IR input-file s
are specified on the command line,
.B html2text
reads from standard input. A dash as the
.I input-file
is an alternate way to specify standard input.
.P
.B html2text
understands all HTML 3.2 constructs, but can render only part of them due to
the limitations of the text output format. However, the program attempts to
provide good substitutes for the elements it cannot render.
.B html2text
parses HTML 4 input, too, but not always as successful as other HTML
processors. It also accepts syntactically incorrect input, and attempts to
interpret it "reasonably".
.P
The way
.B html2text
formats the HTML documents is controlled by formatting properties read
from an RC file.
.B html2text
attempts to read
.I $HOME/.html2textrc
(or the file specified by the
.B -rcfile
command line option); if that file cannot be read,
.B html2text
attempts to read
.IR /etc/html2textrc .
If no RC file can be read (or if the RC file does not override all
formatting properties), then "reasonable" defaults are assumed. The
RC file format is described in the
.BR html2textrc (5)
manual page.
.SH OPTIONS
.TP
.B \-ascii
By default,
.B html2text
uses
.B UTF-8
for its output. Specifying this option, plain
.B ASCII
is used instead. Non\-ASCII characters are rendered via iconv's
transliteration feature.  As such this option is an alias for
.B -to_encoding
.IR ASCII//TRANSLIT .
.TP
.B \-utf8
Assume both terminal and input stream are in
.B UTF-8
mode. This is an alias for
.B -from_encoding
.IR UTF-8
.B -to_encoding
.IR UTF-8 .
.TP
.BI "\-from_encoding " encoding
Sets the encoding of the input file or stream to the given encoding.  By
default,
.B html2text
tries to obtain the encoding from the input file and uses ISO-8859-1
encoding as fallback.  You might want to override the detection or
fallback using this option.  See 
.B iconv -l
for a list of supported encodings.
.TP
.BI "\-to_encoding " encoding
Use the given encoding while writing output, instead of UTF-8.  The same
set of encodings as for the
.B \-from_encoding
option can be used.
.TP
.B \-check
This option is for diagnostic purposes: The HTML document is only parsed and
not processed otherwise. In this mode of operation,
.B html2text
will report on parse errors and scan errors, which it does not in other modes
of operation. Note that parse and scan errors are not fatal for
.BR html2text ,
but may cause mis-interpretation of the HTML code and/or portions of the
document being swallowed.
.TP
.B \-debug\-parser
Let
.B html2text
report on the tokens being shifted, rules being applied, etc., while scanning
the HTML document. This option is for diagnostic purposes.
.TP
.B \-debug\-scanner
Let
.B html2text
report on each lexical token scanned, while scanning the HTML document. This
option is for diagnostic purposes.
.TP
.B \-help
Print command line summary and exit.
.TP
.B \-nobs
By default,
.B html2text
renders underlined letters with sequences like "underscore-backspace-character"
and boldface letters like "character-backspace-character", which works fine
when the output is piped into
.BR more (1),
.BR less (1),
or similar. For other applications, or when redirecting the output into a file,
it may be desirable not to render character attributes with such backspace
sequences, which can be accomplished with this command line option.
.TP
.BI \-o " output\-file"
Write the output to
.I output\-file
instead of standard output. A dash as the
.I output\-file
is an alternate way to specify the standard output.
.TP
.BI \-rcfile " path"
Attempt to read the file specified in
.I path
as RC file.
.TP
.B \-links
Tags all external links with a number between brackets
.RI "([" num "]),"
and produces a numbered list at the end of the document with all link
targets.
.TP
.B \-version
Print program version and exit.
.TP
.BI \-width " width"
By default,
.B html2text
formats the HTML documents for a screen width of 79 characters. If redirecting
the output into a file, or if your terminal has a width other than 80
characters, or if you just want to get an idea how
.B html2text
deals with large tables and different terminal widths, you may want to specify
a different
.IR width .
.SH FILES
.TP
.I /etc/html2textrc
System wide parser configuration file.
.TP
.I $HOME/.html2textrc
Personal parser configuration file, overrides the system wide values.
.SH "CONFORMING TO"
.B "HTML 3.2"
(HTML 3.2 Reference Specification \- http://www.w3.org/TR/REC-html32).
.SH RESTRICTIONS
.B html2text
was written to convert HTML 3.2 documents. When using it with HTML 4 or
even XHTML 1 documents, some constructs present only in these HTML versions
might not be rendered.
.SH AUTHOR
.B html2text
was written up to version 1.2.2 by Arno Unkrig <arno@unkrig.de>
for GMRS Software GmbH, Unterschleissheim.
.P
Up to version 1.3.2a, the maintainer was
Martin Bayer <mbayer@zedat.fu-berlin.de>.
.br
.P
This version of
.B html2text
comes from
https://gitlab.com/grobian/html2text.  Please use the GitLab page to
file issues and improvements.
.SH SEE ALSO
.BR html2textrc (5),
.BR less (1),
.BR more (1)


CHANGES
## CHANGES - What has changed ?                   Wed Jan 14 14:47:26 CET 2004
## ===========================================================================


                         #############################
                         #                           #
                         #Changes in Version 1.3.2(a)#
                         #                           #
                         #   (released 2004-01-15)   #
                         #############################

Bugfixes and new features
****************************************
Version 1.3.2A only: ported to gcc 3.3+

------------------------------
Ported to g++ v. 3.3. Since this change is not compatible to g++ versions
prior to 3.0, it is included in version 1.3.2A only. THIS IS THE ONLY
DIFFERENCE BETWEEN VERSIONS 1.2.3 AND 1.2.3A. (Area.C, Area.h, HTMLControl.C,
HTMLControl.h, Properties.C, Properties.h, format.C, format.h, html.C, html.h,
html2text.C, urlistream.h)
------------------------------
Added command line option '-ascii' for straight ascii output (instead of
ISO-8859-1, which is the default). (html2text.C, sgml.C)
------------------------------
Implemented rendering of most SGML entities introduced in HTML-4. As a
limitation, those entities not present in HTML-3.2/ISO-8859-1 will be
recognized only if represented as "named entities" and not with thier numeric
values, e.g. "&trade;" will be rendered, "&#8482;" won't. (html2text.C,
sgml.C)
------------------------------
Element closing as done in XHTML (e.g. "<br />") is now tolerated.
(HTMLControl.C)
------------------------------
The program now ignores the content of <SCRIPT> or <STYLE> elemts within
TABLEs, even if it is not commented out. (HTMLParser.y)
------------------------------
Fixed urlistream.h (fd_ might be uninitialized). (urlistream.h)
------------------------------


                         #############################
                         #                           #
                         # Changes in Version 1.3.1  #
                         #                           #
                         #   (released 2002-09-02)   #
                         #############################

Bugfix release
****************************************

------------------------------
Multiple-line DOCTYPE declarations are now accepted (HTMLControl.C:352).
------------------------------
Bad initialisation of "fd_" in "urlistream" fixed (urlistream.h), so that an
error message is printed if a remote document could not be retrieved.
------------------------------
A missing node now is interpreted as node '/' (urlistream.C).
------------------------------
Closing DIV, FORM and BLOCKQUOTE tags as well as the closing UL, OL and PRE
tags are now optional (HTMLParser.y), so that the program won't die any longer
if one of them is omited in the document.
------------------------------
Some dificulties with non-ASCII chars fixed (HTMLControl.C).
------------------------------
Colons in elements and attributes now are tolerated and elements are not any
longer implecitly closed at a newline (HTMLControl.C), in order to handly
MS-Word's HTML better.
------------------------------
We do not use '/dev/stdin' as file descriptor any longer (html2text.C and
urlistream.C), in order to make reading from STDIN finaly work.
------------------------------
Block elements are not enclosed in -implecit- Paragraphs any longer
(HTMLParser.y), in order to avoid superfluous newlines in the output.
------------------------------
Fixed segmentation fault on tables with "border" attribute: we now assume that
any TABLE has at least one row and one column (table.C).
------------------------------
Fixed format.C to avoid excessive runtime increment on parsing much nested
block elements.
------------------------------
Some changes in configure, Makefile.in and the documentation.
------------------------------


                         #############################
                         #                           #
                         # Changes in Version 1.3.0  #
                         #                           #
                         #   (released 2001-10-08)   #
                         #############################

Ported to g++-3.0 and GPL
****************************************

------------------------------
Ported to g++ version 3.0.
This uses the 'istream.h' header file from the g++3's 'backward' directory.
------------------------------
Bugfix: '-' did not work as synonym for STDIN.
------------------------------
Added support for the EURO-sign (well, almost).
------------------------------
Finaly the GNU GPL as new copyright terms for all parts of the program, after
GMRS agreed to change the program's license terms to it.
------------------------------


                         #############################
                         #                           #
                         # Changes in Version 1.2.4  #
                         #                           #
                         #   (released 2001-06-11)   #
                         #############################

Bugfix and New Features
****************************************

------------------------------
Fixed coredump when parsing empty <SCRIPT> or <STYLE> elements (HTMLControl.C).
------------------------------
New image handling: <IMG alt=""> does no longer return the value of the SRC
attribute nor "[]". Added new ~/.html2textrc options: IMG.replace.{all noalt}
and IMG.alt.{prefix suffix} with new defaults in pretty-style mode. Added
method for checking whether an attribut was set, even to a zero-value.
------------------------------
(New) Copyright terms for all changes we made since version 1.2.2.
------------------------------
Some minor changes in configure and html2text.C.
------------------------------
Updated the documentation.
------------------------------


                         #############################
                         #                           #

html2text.C

/***************************************************************************/

/*
 * Portions Copyright (c) 1999 GMRS Software GmbH
 * Carl-von-Linde-Str. 38, D-85716 Unterschleissheim, http://www.gmrs.de
 * All rights reserved.
 *
 * Author: Arno Unkrig <arno@unkrig.de>
 */

/* This program is free software; you can redistribute it and/or modify
 * it under the terms of the GNU General Public License as published by
 * the Free Software Foundation; either version 2 of the License, or
 * (at your option) any later version.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the
 * GNU General Public License in the file COPYING for more details.
 */

/***************************************************************************/

/*
 * Changes to version 1.2.2 were made by Martin Bayer <mbayer@zedat.fu-berlin.de>
 * Dates and reasons of modifications:
 * Fre Jun  8 19:00:26 CEST 2001: new image handling
 * Thu Oct  4 21:42:24 CEST 2001: ported to g++ 3.0, bugfix for '-' as synonym for STDIN
 * Mon Jul 22 13:48:26 CEST 2002: Made finaly reading from STDIN work.
 * Sat Sep 14 15:04:09 CEST 2002: Added plain ASCII output patch by Bela Lubkin
 * Wed Jul  2 22:08:45 CEST 2003: ported to g++ 3.3
 */

/***************************************************************************/


#include <iostream>
#include <string.h>
#include <stdlib.h>

#include "html.h"
#include "HTMLControl.h"
#include "urlistream.h"
#include "format.h"

#define stringify(x) stringify2(x)
#define stringify2(x) #x

/* ------------------------------------------------------------------------- */

class MyParser : public HTMLControl {
public:
enum {
	PRINT_AS_ASCII, UNPARSE, SYNTAX_CHECK
};

MyParser(
	urlistream &is_,
	bool debug_scanner_,
	bool debug_parser_,
	ostream    &os_,
	int mode_,
	int width_,
	const char *file_name_
	) :
	HTMLControl(is_, debug_scanner_, debug_parser_),
	os(os_),
	mode(mode_),
	width(width_),
	file_name(file_name_)
{}

private:
/*virtual*/ void yyerror(char *);
/*virtual*/ void process(const Document &);

ostream &os;
int mode;
int width;
string file_name;
};

/*virtual*/ void
MyParser::yyerror(char *p)
{
	/*
	 * Swallow parse error messages if not in "syntax check" mode.
	 */
	if (mode != SYNTAX_CHECK && !strcmp(p, "parse error"))
		return;

	std::cerr
		<< "File \""
		<< file_name
		<< "\", line "
		<< current_line
		<< ", column "
		<< current_column
		<< ": "
		<< p
		<< std::endl;
}

/*virtual*/ void
MyParser::process(const Document &document)
{
	switch (mode) {
	case PRINT_AS_ASCII:
		document.format(/*indent_left*/ 0, width, Area::LEFT, os);
		break;

	case UNPARSE:
		document.unparse(os, std::endl);
		break;

	case SYNTAX_CHECK:
		break;

	default:
		std::cerr << "??? Invalid mode " << mode << " ??? " << std::endl;
		exit(1);
		break;
	}
}

/* ------------------------------------------------------------------------- */

static const char *usage = "\
Usage:\n\
  html2text -help\n\
  html2text -version\n\
  html2text [ -unparse | -check ] [ -debug-scanner ] [ -debug-parser ] \\\n\
     [ -rcfile <file> ] [ -style ( compact | pretty ) ] [ -width <w> ] \\\n\
     [ -o <file> ] [ -nobs ] [ -ascii ] [ <input-url> ] ...\n\
Formats HTML document(s) read from <input-url> or STDIN and generates ASCII\n\
text.\n\
  -help          Print this text and exit\n\
  -version       Print program version and copyright notice\n\
  -unparse       Generate HTML instead of ASCII output\n\
  -check         Do syntax checking only\n\
  -debug-scanner Report parsed tokens on STDERR (debugging)\n\
  -debug-parser  Report parser activity on STDERR (debugging)\n\
  -rcfile <file> Read <file> instead of \"$HOME/.html2textrc\"\n\
  -style compact Create a \"compact\" output format (default)\n\
  -style pretty  Insert some vertical space for nicer output\n\
  -width <w>     Optimize for screen widths other than 79\n\
  -o <file>      Redirect output into <file>\n\
  -nobs          Do not use backspaces for boldface and underlining\n\
  -ascii         Use plain ASCII for output instead of ISO-8859-1\n\
  -utf8          Assume both terminal and input stream are in UTF-8 mode\n\
";

int use_encoding = ISO8859;

int
main(int argc, char **argv)
{
	if (argc == 2 && !strcmp(argv[1], "-help")) {
		std::cout
			<< "This is html2text, version " stringify(VERSION) << std::endl
			<< std::endl
			<< usage;
		exit(0);
	}

	if (argc == 2 && !strcmp(argv[1], "-version")) {
		std::cout
			<< "This is html2text, version " stringify(VERSION) << std::endl
			<< std::endl
			<< "The latest version can be found at http://userpage.fu-berlin.de/~mbayer/tools/" << std::endl
			<< std::endl
			<< "This program is distributed in the hope that it will be useful, but WITHOUT" << std::endl
			<< "ANY WARRANTY; without even the implied warranty of MERCHANTABILITY or FITNESS" << std::endl
			<< "FOR A PARTICULAR PURPOSE. See the GNU General Public License for more details." << std::endl
			<< std::endl;
		exit(0);
	}

	bool mode = MyParser::PRINT_AS_ASCII;
	bool debug_scanner = false;
	bool debug_parser = false;
	const char *home = getenv("HOME");
	string rcfile = string(home ? home : "") + "/.html2textrc";
	const char *style = "compact";
	int width = 79;
	const char *output_file_name = "-";
	bool use_backspaces = true;

	int i;
	for (i = 1; i < argc && argv[i][0] == '-' && argv[i][1]; i++) {
		const char *arg = argv[i];

		if (!strcmp(arg, "-unparse")) {
			mode = MyParser::UNPARSE;
		} else if (!strcmp(arg, "-check")) {
			mode = MyParser::SYNTAX_CHECK;
		} else if (!strcmp(arg, "-debug-scanner")) {
			debug_scanner = true;
		} else if (!strcmp(arg, "-debug-parser")) {
			debug_parser = true;
		} else if (!strcmp(arg, "-rcfile")) {
			rcfile = argv[++i];
		} else if (!strcmp(arg, "-style")) {
			style = argv[++i];
		} else if (!strcmp(arg, "-width")) {
			width = atoi(argv[++i]);
		} else if (!strcmp(arg, "-o")) {
			output_file_name = argv[++i];
		} else if (!strcmp(arg, "-nobs")) {
			use_backspaces = false;
		} else if (!strcmp(arg, "-ascii")) {
			use_encoding = ASCII;
		} else if (!strcmp(arg, "-utf8")) {
			use_encoding = UTF8;
		} else {
			std::cerr
				<< "Unrecognized command line option \""
				<< arg
				<< "\", try \"-help\"."
				<< std::endl;
			exit(1);
		}
	}
	if (i > argc) {
		std::cerr
			<< "Error: Required parameter after \""
			<< argv[argc - 1]
			<< "\" missing."
			<< std::endl;
		exit(1);
	}

	const char *const *input_urls;
	int number_of_input_urls;

	if (i >= argc) {
		static const char *const x = "-";
		input_urls = &x;
		number_of_input_urls = 1;
	} else {
		input_urls = argv + i;
		number_of_input_urls = argc - i;
	}

	/*
	 * Set up formatting: First, set some formatting properties depending on
	 * the "-style" command line option.
	 */
	if (!strcmp(style, "compact")) {
		;
	} else if (!strcmp(style, "pretty")) {
		/*
		 * The "pretty" style was kindly supplied by diligent user Rolf Niepraschk.
		 */
		static const struct {
			const char *key;
			const char *value;
		}
		properties[] = {
			{ "OL.TYPE", "1" },
			{ "OL.vspace.before", "1" },
			{ "OL.vspace.after", "1" },
			{ "OL.indents", "5" },
			{ "UL.vspace.before", "1" },
			{ "UL.vspace.after", "1" },
			{ "UL.indents", "2" },
			{ "DL.vspace.before", "1" },
			{ "DL.vspace.after", "1" },
			{ "DT.vspace.before", "1" },
			{ "DIR.vspace.before", "1" },
			{ "DIR.indents", "2" },
			{ "MENU.vspace.before", "1" },
			{ "MENU.vspace.after", "1" },
			{ "DT.indent", "2" },
			{ "DD.indent", "6" },
			{ "HR.marker", "-" },
			{ "H1.prefix", "" },
			{ "H2.prefix", "" },
			{ "H3.prefix", "" },
			{ "H4.prefix", "" },
			{ "H5.prefix", "" },
			{ "H6.prefix", "" },
			{ "H1.suffix", "" },
			{ "H2.suffix", "" },
			{ "H3.suffix", "" },
			{ "H4.suffix", "" },
			{ "H5.suffix", "" },
			{ "H6.suffix", "" },
			{ "H1.vspace.before", "2" },
			{ "H2.vspace.before", "1" },
			{ "H3.vspace.before", "1" },
			{ "H4.vspace.before", "1" },
			{ "H5.vspace.before", "1" },
			{ "H6.vspace.before", "1" },
			{ "H1.vspace.after", "1" },
			{ "H2.vspace.after", "1" },
			{ "H3.vspace.after", "1" },
			{ "H4.vspace.after", "1" },
			{ "H5.vspace.after", "1" },
			{ "H6.vspace.after", "1" },
			{ "TABLE.vspace.before", "1" },
			{ "TABLE.vspace.after", "1" },
			{ "CODE.vspace.before", "0" },
			{ "CODE.vspace.after", "0" },
			{ "BLOCKQUOTE.vspace.before", "1" },
			{ "BLOCKQUOTE.vspace.after", "1" },
			{ "PRE.vspace.before", "1" },
			{ "PRE.vspace.after", "1" },
			{ "PRE.indent.left", "2" },
			{ "IMG.replace.noalt", "" },
			{ "IMG.alt.prefix", " " },
			{ "IMG.alt.suffix", " " },
			{ 0, 0 }
		}, *p;
		for (p = properties; p->key; ++p) {
			Formatting::setProperty(p->key, p->value);
		}
	} else {
		std::cerr
			<< "Unknown style \""
			<< style
			<< "\" specified -- try \"-help\"."
			<< std::endl;
		::exit(1);
	}

	{
		std::ifstream ifs(rcfile.c_str());
		if (!ifs.rdbuf()->is_open())
			ifs.open("/etc/html2textrc");
		if (ifs.rdbuf()->is_open()) {
			Formatting::loadProperties(ifs);
		}
	}

	/*
	 * Set up printing.
	 */
	Area::use_backspaces = use_backspaces;

	ostream  *osp;
	std::ofstream ofs;

	if (!strcmp(output_file_name, "-")) {
		osp = &std::cout;
	} else {
		ofs.open(output_file_name, std::ios::out);
		if (!ofs) {
			std::cerr
				<< "Could not open output file \""
				<< output_file_name
				<< "\"."
				<< std::endl;
			exit(1);
		}
		osp = &ofs;
	}

	for (i = 0; i < number_of_input_urls; ++i) {
		const char *input_url = input_urls[i];

		if (number_of_input_urls != 1) {
			*osp << "###### " << input_url << " ######" << std::endl;
		}

		istream    *isp;
		urlistream uis;

		uis.open(input_url);
		if (!uis.is_open()) {
			std::cerr
				<< "Opening input URL \""
				<< input_url
				<< "\": "
				<< uis.open_error()
				<< std::endl;
			exit(1);
		}

		MyParser parser(
			uis,
			debug_scanner,
			debug_parser,
			*osp,
			mode,
			width,
			input_url
			);

		if (parser.yyparse() != 0)
			exit(1);
	}

	return 0;
}

/* ------------------------------------------------------------------------- */
